### ISSUE TITLE
Dependency injection and class inheritance doesn't work in Angular 8 library test

### ISSUE THREAD

**Issue by ddramone** [angular/angular#31337]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core@^8

### Is this a regression?
Yes, the previous version in which this bug was not present was: @angular/core@7

### Description

The problem occurs when:

1. The base class has a constructor with an injected service. (`Injector` from the example)

```ts
@Injectable()
export class BaseService {
  // "Injector" is just for the sake of example. Any injectable would have the same problem.
  constructor(public injector: Injector) {
  }
}
```
2. Child class has no constructor (i.e., it relies on base class constructor) AND at least one property with the default value.

```ts
@Injectable()
export class MyLibService extends BaseService {

  public foo = 'bar';

}
```

3. (_dont ask me how I have found this, but:_) In `angular.json` `[lib name].test.options.codeCoverage` should be true. 

### PROBLEM:   

`injector` is undefined in the instance of `MyLibService`. 

--- 

🚩 Removing property from the child class _(2)_ AND/OR opting out `codeCoverage` config in angular.json _(3)_ doesn't brake test.

As a result, I can't have a class inheritance and code coverage at the same time.

---

I could put a constructor in the child class and call `super` by passing all required arguments manually, but it's the last option since the real-life scenario is much more complicated than the example provided here.

## 🔬 Minimal Reproduction
git repo: https://github.com/ddramone/ng-8-library-inheritance-di-test-bug

reproduce steps:

1. Install angular 8 cli
2. Clone `git clone https://github.com/ddramone/ng-8-library-inheritance-di-test-bug.git `
3. `npm i`
4. run Test `ng test my-lib`

 ### Expected behavior

Test should pass = Injector service should be available in the service instance.

**Comment by Airblader** [comment]
Related to https://github.com/Microsoft/TypeScript/issues/6979 I think.

**Comment by alxhub** [comment]
Possibly related to #31333? @ddramone can you try again with the latest Angular/CLI?

**Comment by ddramone** [comment]
I've tried the latest cli (8.1.0), didn't work. (updated package versions in the repo as well)

**Comment by richardlnnr** [comment]
I've had the same problem

**Comment by andreyjamer** [comment]
I got the same problem. 
It appears only if tests are running with code coverage: `ng test --code-coverage`
The usual `ng test` or any type of build (JIT or AOT) gives no errors.

For now the problem can be solved by injecting services on child level and passing them to parent component.

**Comment by petrce** [comment]
tested on angular 8.2.11 with cli 8.3.12 and it is still not fixed, now I have to add tons of workarounds :(

**Comment by CodeAndWeb** [comment]
Same here. It seems to be somewhat random when it happens.

I have 2 classes using the same base class. The test always fails for one derived class and works for the other.

I can also confirm that it only fails if "--code-coverage" is used.

**Comment by jpzwarte** [comment]
Same problem. 

```
Angular CLI: 8.3.18
Node: 12.12.0
OS: darwin x64
Angular: 8.2.13
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.803.18
@angular-devkit/build-angular      0.803.18
@angular-devkit/build-ng-packagr   0.803.18
@angular-devkit/build-optimizer    0.803.18
@angular-devkit/build-webpack      0.803.18
@angular-devkit/core               8.3.18
@angular-devkit/schematics         8.3.18
@angular/cdk                       8.2.3
@angular/cdk-experimental          8.2.3
@angular/cli                       8.3.18
@ngtools/webpack                   8.3.18
@schematics/angular                8.3.18
@schematics/update                 0.803.18
ng-packagr                         5.7.1
rxjs                               6.5.3
typescript                         3.5.3
webpack                            4.39.2
```

**Comment by jollsker** [comment]
Same problem here


`Angular CLI: 8.3.4
Node: 10.16.3
OS: win32 x64
Angular: 8.2.6
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.803.4
@angular-devkit/build-angular      0.803.4
@angular-devkit/build-ng-packagr   0.803.4
@angular-devkit/build-optimizer    0.803.4
@angular-devkit/build-webpack      0.803.4
@angular-devkit/core               8.3.4
@angular-devkit/schematics         8.3.4
@angular/cdk                       8.2.0
@angular/cli                       8.3.4
@angular/material                  8.2.0
@ngtools/webpack                   8.3.4
@schematics/angular                8.3.4
@schematics/update                 0.803.4
ng-packagr                         5.5.1
rxjs                               6.5.2
typescript                         3.5.3
webpack                            4.39.2`

**Comment by sbjw** [comment]
i have the same problem.
this is my temp solve method
when serve:
modify the packge.json  to
        "@angular-devkit/build-angular": "~0.802.2",(i try all 0.803.*, it don't work)
when build:
modify the packge.json  to
        "@angular-devkit/build-angular": "~0.803.19",



this is my environment

Angular CLI: 8.3.19
Node: 12.13.1
OS: win32 x64
Angular:
...

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.803.19
@angular-devkit/core         8.3.19
@angular-devkit/schematics   8.3.19
@schematics/angular          8.3.19
@schematics/update           0.803.19
rxjs                         6.4.0

**Comment by mlakmal** [comment]
we are seeing the same issue when using differential loading with v8 and this is breaking in es5 script bundles in IE 11 for specific instances where we have multiple class inheritances without constructor defined in some of the classes.

ex:
`
export class A {
constructor(public dep1: SomeType){}
}

export class B extends A {
}

export class C extends B {
constructor(dep1: SomeType, dep2: AnotherType){
super(dep1);
}

ngOnInit(){
//this.dep1 is undefined here
//this.dep2 is defined
}
}
`

**Comment by jonathanlie** [comment]
bumping, the exact same problem, but specifically only in IE11.

We have a component that extends an abstract class that injects `ComponentFactoryResolver` (to resolve dynamic components), and it worked before in IE11; now the dependency is `undefined` in the component.

```
abstract class WrapperComponent {
  constructor(
    private componentFactoryResolver: ComponentFactoryResolver
  ) {
  }
}

class AComponent extends WrapperComponent {
  
}
```

**Comment by KostyaTretyak** [comment]
By the way, if you use `Injector` directly, DI works as it should (with `@angular/core` v8.2.14).

```ts
import { Injector, Injectable } from '@angular/core';

class Dependency {
  prop1 = 'this is dependency property';
}

@Injectable()
class Parent {
  constructor(public dependency: Dependency) {}
}

@Injectable()
class Child extends Parent { }

const injector = Injector.create({providers: [
  {provide: Dependency, useClass: Dependency, deps: []},
  {provide: Parent, useClass: Parent, deps: []},
  {provide: Child, useClass: Child, deps: [Dependency]}
  ]});

console.log(injector.get(Child));
```

See [stackblitz](https://stackblitz.com/edit/angular-angular-issues-31337?file=index.ts)

**Comment by havran** [comment]
Same problem here:

```
Angular CLI: 9.0.0-rc.8
Node: 10.16.3
OS: linux x64

Angular: 9.0.0-rc.8
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router
Ivy Workspace: No

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.0-rc.8
@angular-devkit/build-angular      0.900.0-rc.8
@angular-devkit/build-ng-packagr   0.900.0-rc.9
@angular-devkit/build-optimizer    0.900.0-rc.8
@angular-devkit/build-webpack      0.900.0-rc.8
@angular-devkit/core               9.0.0-rc.8
@angular-devkit/schematics         9.0.0-rc.8
@angular/language-service          9.0.0-rc.9
@ngtools/webpack                   9.0.0-rc.8
@nguniversal/builders              9.0.0-rc.0
@nguniversal/common                9.0.0-rc.0
@nguniversal/express-engine        9.0.0-rc.0
@schematics/angular                9.0.0-rc.8
@schematics/update                 0.900.0-rc.8
ng-packagr                         9.0.0-rc.7
rxjs                               6.5.4
typescript                         3.6.4
webpack                            4.41.2
```

**Comment by zstundys** [comment]
Workaround that worked for me:

- changing the `compilerOptions.target` property to `es5` in `tsconfig.json` 

Although it fixes the `--code-coverage` test runs, it prevents differential (es5, es2015) builds, though - resulting in bigger bundles

**Comment by JoostK** [comment]
I have looked into why DI doesn't work when coverage reporting is enabled, and it can be traced back to how Angular detects whether a constructor was present in a class, or if a constructor is generated synthetically, e.g. to initialize class fields. Enabling coverage inserts coverage instrumentation statements into the generated constructors:

```js
class MyLibService extends BaseService {
  constructor() {
    cov_8nt6qq5zt().f[2]++;
    cov_8nt6qq5zt().s[6]++;
    super(...arguments);
    cov_8nt6qq5zt().s[7]++;
    this.foo = 'bar';
  }
}
```

These prevent Angular from recognizing the constructor as synthesized, causing `MyLibService` to be instantiated with 0 arguments. The proper behavior would be to delegate to the parent constructor, allowing its dependencies to be injected.

I opened #37811 as a proposal to workaround this issue. It's not done yet and I can't give an ETA yet, unfortunately.
